File: src/main/java/com/github/mbuzdalov/opl/cma/FitnessHistory.java
package com.github.mbuzdalov.opl.cma;

public final class FitnessHistory {
    private final double[] inputStack, outputMin, outputMax;
    private int nInputs, nOutputs;
    private double inputMin, inputMax;

    public FitnessHistory(int length) {
        inputStack = new double[length];
        outputMin = new double[length];
        outputMax = new double[length];
        inputMin = Double.POSITIVE_INFINITY;
        inputMax = Double.NEGATIVE_INFINITY;
    }

    public void push(double value) {
        if (nInputs + nOutputs == getCapacity()) {
            if (nOutputs == 0) {
                double tailMin = Double.POSITIVE_INFINITY;
                double tailMax = Double.NEGATIVE_INFINITY;
                while (nInputs > 0) {
                    --nInputs;
                    double popped = inputStack[nInputs];
                    tailMin = Math.min(tailMin, popped);
                    tailMax = Math.max(tailMax, popped);
                    outputMin[nOutputs] = tailMin;
                    outputMax[nOutputs] = tailMax;
                    ++nOutputs;
                }
                inputMin = Double.POSITIVE_INFINITY;
                inputMax = Double.NEGATIVE_INFINITY;
            }
            --nOutputs;
        }
        inputStack[nInputs] = value;
        ++nInputs;
        inputMin = Math.min(inputMin, value);
        inputMax = Math.max(inputMax, value);
    }

    public int getCapacity() {
        return inputStack.length;
    }

    public double getMinimum() {
        return nOutputs == 0 ? inputMin : Math.min(inputMin, outputMin[nOutputs - 1]);
    }

    public double getMaximum() {
        return nOutputs == 0 ? inputMax : Math.max(inputMax, outputMax[nOutputs - 1]);
    }
}

File: src/main/scala/com/github/mbuzdalov/util/FitnessHistory.scala
package com.github.mbuzdalov.util

class FitnessHistory(length: Int):
  private val inputStack, outputMin, outputMax = Array.ofDim[Double](length)
  private var nInputs, nOutputs = 0
  private var inputMin = Double.PositiveInfinity
  private var inputMax = Double.NegativeInfinity

  def push(value: Double): Unit =
    if nInputs + nOutputs == capacity then
      if nOutputs == 0 then
        var tailMin = Double.PositiveInfinity
        var tailMax = Double.NegativeInfinity
        while nInputs > 0 do
          nInputs -= 1
          val popped = inputStack(nInputs)
          tailMin = math.min(tailMin, popped)
          tailMax = math.max(tailMax, popped)
          outputMin(nOutputs) = tailMin
          outputMax(nOutputs) = tailMax
          nOutputs += 1
        inputMin = Double.PositiveInfinity
        inputMax = Double.NegativeInfinity
      nOutputs -= 1
    inputStack(nInputs) = value
    nInputs += 1
    inputMin = math.min(inputMin, value)
    inputMax = math.max(inputMax, value)

  def capacity: Int = inputStack.length
  def minimum: Double = if nOutputs == 0 then inputMin else math.min(inputMin, outputMin(nOutputs - 1))
  def maximum: Double = if nOutputs == 0 then inputMax else math.max(inputMax, outputMax(nOutputs - 1))
